File: dotnet/PolymarketBot/Services/Portfolio.cs
using Microsoft.Extensions.Logging;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

public sealed class Portfolio
{
    private const double InputCostPerMTok = 3.0;
    private const double OutputCostPerMTok = 15.0;

    private readonly BotConfig _config;
    private readonly ILogger<Portfolio> _log;

    public double Bankroll { get; private set; }
    public double InitialBankroll { get; private set; }
    public List<Position> Positions { get; private set; }
    public double HighWaterMark { get; private set; }
    public double DailyStartValue { get; private set; }
    public double TotalRealizedPnl { get; private set; }
    public int TotalTrades { get; private set; }
    public bool IsHalted { get; set; }
    public double TotalApiCost { get; private set; }

    public Portfolio(BotConfig config, ILogger<Portfolio> log, PortfolioSnapshot? snapshot = null)
    {
        _config = config;
        _log = log;

        if (snapshot is not null)
        {
            Bankroll = snapshot.Bankroll;
            InitialBankroll = snapshot.InitialBankroll;
            Positions = new List<Position>(snapshot.Positions);
            HighWaterMark = snapshot.HighWaterMark;
            DailyStartValue = snapshot.DailyStartValue;
            TotalRealizedPnl = snapshot.TotalRealizedPnl;
            TotalTrades = snapshot.TotalTrades;
            IsHalted = snapshot.IsHalted;
        }
        else
        {
            Bankroll = config.InitialBankroll;
            InitialBankroll = config.InitialBankroll;
            Positions = [];
            HighWaterMark = config.InitialBankroll;
            DailyStartValue = config.InitialBankroll;
            TotalRealizedPnl = 0;
            TotalTrades = 0;
            IsHalted = false;
        }

        TotalApiCost = 0;
    }

    public PortfolioSnapshot Snapshot() => new()
    {
        Bankroll = Bankroll,
        InitialBankroll = InitialBankroll,
        Positions = new List<Position>(Positions),
        HighWaterMark = HighWaterMark,
        DailyStartValue = DailyStartValue,
        TotalRealizedPnl = TotalRealizedPnl,
        TotalTrades = TotalTrades,
        IsHalted = IsHalted,
    };

    public double TotalExposure() => Positions.Sum(p => p.SizeUsd);

    public double CategoryExposure(string category)
        => Positions.Where(p => p.Category == category).Sum(p => p.SizeUsd);

    public bool HasPosition(string conditionId)
        => Positions.Any(p => p.ConditionId == conditionId);

    // -- Signal generation --

    public Signal? GenerateSignal(MarketInfo market, Estimate estimate)
    {
        var fair = estimate.FairProbability;
        var yesEdge = fair - market.OutcomeYesPrice;
        var noEdge = (1.0 - fair) - market.OutcomeNoPrice;

        Side side;
        double edge, marketPrice;

        if (yesEdge > noEdge && yesEdge > _config.MinEdge)
        {
            side = Side.YES;
            edge = yesEdge;
            marketPrice = market.OutcomeYesPrice;
        }
        else if (noEdge > _config.MinEdge)
        {
            side = Side.NO;
            edge = noEdge;
            marketPrice = market.OutcomeNoPrice;
        }
        else
        {
            return null;
        }

        if (marketPrice <= 0 || marketPrice >= 1) return null;

        // Kelly criterion: f* = (b*p - q) / b
        var b = (1.0 / marketPrice) - 1.0;
        var p = side == Side.YES ? fair : 1.0 - fair;
        var q = 1.0 - p;
        var kellyRaw = b > 0 ? (b * p - q) / b : 0.0;
        kellyRaw = Math.Max(0.0, kellyRaw);

        // Fractional Kelly + position cap (use portfolio value, not just cash)
        var kelly = kellyRaw * _config.KellyFraction;
        var portfolioVal = Bankroll + TotalExposure();
        var sizeUsd = kelly * portfolioVal;
        sizeUsd = Math.Min(sizeUsd, portfolioVal * _config.MaxPositionPct);
        sizeUsd = Math.Min(sizeUsd, Bankroll); // never exceed available cash

        if (sizeUsd < _config.MinTradeUsd) return null;

        // CLOB minimum order size is 5 tokens → minimum USD = 5 * price
        var minClobUsd = 5.0 * marketPrice;
        if (sizeUsd < minClobUsd)
        {
            _log.LogInformation("Position ${Size:F2} below CLOB minimum ${Min:F2} (5 tokens @ {Price:F3})",
                sizeUsd, minClobUsd, marketPrice);
            return null;
        }

        return new Signal
        {
            Market = market,
            Estimate = estimate,
            Side = side,
            Edge = edge,
            MarketPrice = marketPrice,
            KellyFraction = kelly,
            PositionSizeUsd = Math.Round(sizeUsd, 2),
            ExpectedValue = Math.Round(sizeUsd * edge, 4),
        };
    }

    // -- Risk checks --

    public bool CheckRisk(Signal signal)
    {
        if (HasPosition(signal.Market.ConditionId))
        {
            _log.LogInformation("Risk BLOCK: already positioned in {Question}", Truncate(signal.Market.Question, 40));
            return false;
        }

        if (Positions.Count >= _config.MaxConcurrentPositions)
        {
            _log.LogInformation("Risk BLOCK: max positions ({Max}) reached", _config.MaxConcurrentPositions);
            return false;
        }

        var pv = Bankroll + TotalExposure();
        var newExposure = TotalExposure() + signal.PositionSizeUsd;
        var maxAllowed = pv * _config.MaxTotalExposurePct;
        if (newExposure > maxAllowed)
        {
            _log.LogInformation("Risk BLOCK: total exposure ${New:F2} > limit ${Limit:F2}", newExposure, maxAllowed);
            return false;
        }

        var catExp = CategoryExposure(signal.Market.Category) + signal.PositionSizeUsd;
        var catLimit = pv * _config.MaxCategoryExposurePct;
        if (catExp > catLimit)
        {
            _log.LogInformation("Risk BLOCK: '{Category}' exposure ${Exp:F2} > limit ${Limit:F2}",
                signal.Market.Category, catExp, catLimit);
            return false;
        }

        // Daily stop loss (include open position value — deployed capital isn't lost)
        var portfolioValue = Bankroll + TotalExposure();
        var dailyPnl = portfolioValue - DailyStartValue;
        if (dailyPnl < 0 && Math.Abs(dailyPnl) > DailyStartValue * _config.DailyStopLossPct)
        {
            _log.LogWarning("HALT: Daily stop loss triggered (PnL=${Pnl:+0.00;-0.00}, limit={Limit:P0})",
                dailyPnl, _config.DailyStopLossPct);
            IsHalted = true;
            return false;
        }

        // Max drawdown from high water mark
        if (HighWaterMark > 0)
        {
            var drawdown = (HighWaterMark - portfolioValue) / HighWaterMark;
            if (drawdown > _config.MaxDrawdownPct)
            {
                _log.LogWarning("HALT: Max drawdown {Drawdown:P1} exceeded (limit={Limit:P0})",
                    drawdown, _config.MaxDrawdownPct);
                IsHalted = true;
                return false;
            }
        }

        // Agent death — only when total portfolio value (free cash + open positions)
        // drops below $1. Negative bankroll from API costs while holding positions
        // is normal: positions will eventually resolve and return USDC.
        if (Bankroll + TotalExposure() < 1.0)
        {
            _log.LogWarning("HALT: Portfolio value < $1 — agent is dead");
            IsHalted = true;
            return false;
        }

        return true;
    }

    // -- Position management --

    public void OpenPosition(Position position)
    {
        Bankroll -= position.SizeUsd;
        TotalTrades++;
        Positions.Add(position);
        _log.LogInformation("Opened {Side} on {Question} ${Size:F2} @ {Price:F3}",
            position.Side, Truncate(position.Question, 40), position.SizeUsd, position.EntryPrice);
    }

    public double ClosePosition(string conditionId, double exitPrice)
    {
        var pos = Positions.FirstOrDefault(p => p.ConditionId == conditionId);
        if (pos is null) return 0.0;

        var pnl = pos.Shares * (exitPrice - pos.EntryPrice);
        Bankroll += pos.SizeUsd + pnl;
        TotalRealizedPnl += pnl;
        Positions = Positions.Where(p => p.ConditionId != conditionId).ToList();
        HighWaterMark = Math.Max(HighWaterMark, Bankroll);

        _log.LogInformation("Closed {Question} PnL: ${Pnl:+0.00;-0.00}", Truncate(pos.Question, 40), pnl);
        return pnl;
    }

    public double ResolvePosition(string conditionId, bool won)
    {
        var pos = Positions.FirstOrDefault(p => p.ConditionId == conditionId);
        if (pos is null) return 0.0;

        var payout = won ? pos.Shares : 0.0;
        var pnl = payout - pos.SizeUsd;
        Bankroll += payout;
        TotalRealizedPnl += pnl;
        TotalTrades++;
        Positions = Positions.Where(p => p.ConditionId != conditionId).ToList();
        HighWaterMark = Math.Max(HighWaterMark, Bankroll);

        var result = won ? "WON" : "LOST";
        _log.LogInformation("Resolved ({Result}): {Question} payout=${Payout:F2}, PnL=${Pnl:+0.00;-0.00}",
            result, Truncate(pos.Question, 40), payout, pnl);
        return pnl;
    }

    // -- Position review --

    public void UpdatePositionPrices(Dictionary<string, double> prices)
    {
        foreach (var pos in Positions)
        {
            if (prices.TryGetValue(pos.TokenId, out var price))
            {
                pos.CurrentPrice = price;
                pos.UnrealizedPnl = pos.Shares * (pos.CurrentPrice - pos.EntryPrice);
            }
        }
    }

    public List<ExitSignal> GenerateExitSignals()
    {
        var signals = new List<ExitSignal>();
        foreach (var pos in Positions)
        {
            // Skip unsellable positions: penny prices or below CLOB minimum (5 tokens)
            if (pos.CurrentPrice < 0.01)
            {
                _log.LogDebug("Skip review: {Question} (price {Price:F4} < $0.01)",
                    Truncate(pos.Question, 40), pos.CurrentPrice);
                continue;
            }
            if (pos.Shares < 5.0)
            {
                _log.LogDebug("Skip review: {Question} ({Shares:F2} tokens < 5 minimum)",
                    Truncate(pos.Question, 40), pos.Shares);
                continue;
            }

            var pnl = pos.Shares * (pos.CurrentPrice - pos.EntryPrice);
            var pnlPct = pos.EntryPrice > 0 ? (pos.CurrentPrice - pos.EntryPrice) / pos.EntryPrice : 0.0;

            // Stop-loss
            if (pnlPct < -_config.PositionStopLossPct)
            {
                signals.Add(new ExitSignal { Position = pos, ExitReason = "stop_loss", CurrentPrice = pos.CurrentPrice, UnrealizedPnl = pnl, PnlPct = pnlPct });
                continue;
            }

            // Take-profit
            if (pos.CurrentPrice >= _config.TakeProfitPrice)
            {
                signals.Add(new ExitSignal { Position = pos, ExitReason = "take_profit", CurrentPrice = pos.CurrentPrice, UnrealizedPnl = pnl, PnlPct = pnlPct });
                continue;
            }

            // Edge-gone
            if (pos.FairEstimateAtEntry > 0)
            {
                var fairForSide = pos.Side == Side.YES ? pos.FairEstimateAtEntry : 1.0 - pos.FairEstimateAtEntry;
                if (pos.CurrentPrice > fairForSide + _config.ExitEdgeBuffer)
                {
                    signals.Add(new ExitSignal { Position = pos, ExitReason = "edge_gone", CurrentPrice = pos.CurrentPrice, UnrealizedPnl = pnl, PnlPct = pnlPct });
                    continue;
                }
            }
        }
        return signals;
    }

    public List<TopupCandidate> GenerateTopupCandidates()
    {
        var candidates = new List<TopupCandidate>();
        foreach (var pos in Positions)
        {
            if (pos.CurrentPrice < 0.01) continue;  // penny = unsellable even with top-up
            if (pos.Shares >= 5.0) continue;         // can sell normally
            if (pos.EntryPrice <= 0) continue;

            var pnlPct = (pos.CurrentPrice - pos.EntryPrice) / pos.EntryPrice;

            // Check same exit conditions as GenerateExitSignals
            string? exitReason = null;
            if (pnlPct < -_config.PositionStopLossPct)
                exitReason = "stop_loss";
            else if (pos.CurrentPrice >= _config.TakeProfitPrice)
                exitReason = "take_profit";
            else if (pos.FairEstimateAtEntry > 0)
            {
                var fairForSide = pos.Side == Side.YES ? pos.FairEstimateAtEntry : 1.0 - pos.FairEstimateAtEntry;
                if (pos.CurrentPrice > fairForSide + _config.ExitEdgeBuffer)
                    exitReason = "edge_gone";
            }

            if (exitReason is null) continue;

            var topupCost = 5.0 * pos.CurrentPrice;
            var recoveryValue = pos.Shares * pos.CurrentPrice;

            candidates.Add(new TopupCandidate
            {
                Position = pos,
                ExitReason = exitReason,
                TokensToBuy = 5.0,
                TopupCost = topupCost,
                RecoveryValue = recoveryValue,
            });
        }
        return candidates;
    }

    public void AddToPosition(string conditionId, double additionalShares, double additionalCost)
    {
        var pos = Positions.FirstOrDefault(p => p.ConditionId == conditionId);
        if (pos is null) return;
        pos.Shares += additionalShares;
        pos.SizeUsd += additionalCost;
        Bankroll -= additionalCost;
        _log.LogInformation(
            "Top-up: +{Shares:F2} tokens (${Cost:F2}) -> {Question} now {Total:F2} tokens",
            additionalShares, additionalCost, Truncate(pos.Question, 40), pos.Shares);
    }

    public List<Position> GetReviewCandidates()
    {
        var candidates = new List<Position>();
        foreach (var pos in Positions)
        {
            if (pos.EntryPrice <= 0) continue;
            var priceMove = Math.Abs(pos.CurrentPrice - pos.EntryPrice) / pos.EntryPrice;
            if (priceMove >= _config.ReviewReestimateThresholdPct)
                candidates.Add(pos);
        }
        candidates.Sort((a, b) => b.SizeUsd.CompareTo(a.SizeUsd));
        return candidates;
    }

    // -- Balance sync --

    /// <summary>
    /// Sync bankroll from actual on-chain USDC balance.
    /// On-chain USDC is always the free cash (bankroll) — conditional tokens
    /// are held separately in the CLOB and not reflected in USDC balance.
    /// Always syncs both up and down so the bot has an accurate view of
    /// spendable funds (handles resolved-position payouts, deposits, fees, etc.)
    /// </summary>
    public void SyncBalance(double actualUsdcBalance)
    {
        var prevBankroll = Bankroll;
        var diff = actualUsdcBalance - prevBankroll;
        if (Math.Abs(diff) <= 0.001)
            return;

        Bankroll = actualUsdcBalance;
        if (diff > 0)
            _log.LogInformation(
                "Balance sync (upward): ${Old:F2} -> ${New:F2} (+${Diff:F2}, {Count} positions open)",
                prevBankroll, Bankroll, diff, Positions.Count);
        else
            _log.LogWarning(
                "Balance sync (downward): ${Old:F2} -> ${New:F2} (${Diff:F2}, {Count} positions open)",
                prevBankroll, Bankroll, diff, Positions.Count);

        HighWaterMark = Math.Max(HighWaterMark, Bankroll + TotalExposure());
    }

    // -- Cost tracking --

    public void RecordApiCost(int inputTokens, int outputTokens)
    {
        var cost = (inputTokens * InputCostPerMTok / 1_000_000.0) +
                   (outputTokens * OutputCostPerMTok / 1_000_000.0);
        TotalApiCost += cost;
        Bankroll -= cost;
    }

    public void ResetDaily()
    {
        DailyStartValue = Bankroll;
    }

    private static string Truncate(string s, int maxLen)
        => s.Length <= maxLen ? s : s[..maxLen] + "...";
}

File: python/portfolio.py
"""Portfolio management, risk checks, and Kelly criterion position sizing."""

import logging
from typing import Optional

from config import BotConfig
from models import MarketInfo, Estimate, Signal, Side, Position, PortfolioSnapshot, ExitSignal, TopupCandidate

log = logging.getLogger("bot.portfolio")

# Approximate Claude API pricing (per million tokens)
INPUT_COST_PER_MTOK = 3.0
OUTPUT_COST_PER_MTOK = 15.0


class Portfolio:
    def __init__(self, config: BotConfig, snapshot: Optional[PortfolioSnapshot] = None):
        self.config = config
        if snapshot:
            self.bankroll = snapshot.bankroll
            self.initial_bankroll = snapshot.initial_bankroll
            self.positions = list(snapshot.positions)
            self.high_water_mark = snapshot.high_water_mark
            self.daily_start_value = snapshot.daily_start_value
            self.total_realized_pnl = snapshot.total_realized_pnl
            self.total_trades = snapshot.total_trades
            self.is_halted = snapshot.is_halted
        else:
            self.bankroll = config.initial_bankroll
            self.initial_bankroll = config.initial_bankroll
            self.positions = []
            self.high_water_mark = config.initial_bankroll
            self.daily_start_value = config.initial_bankroll
            self.total_realized_pnl = 0.0
            self.total_trades = 0
            self.is_halted = False

        self.total_api_cost = 0.0

    def snapshot(self) -> PortfolioSnapshot:
        return PortfolioSnapshot(
            bankroll=self.bankroll,
            initial_bankroll=self.initial_bankroll,
            positions=list(self.positions),
            high_water_mark=self.high_water_mark,
            daily_start_value=self.daily_start_value,
            total_realized_pnl=self.total_realized_pnl,
            total_trades=self.total_trades,
            is_halted=self.is_halted,
        )

    def total_exposure(self) -> float:
        return sum(p.size_usd for p in self.positions)

    def category_exposure(self, category: str) -> float:
        return sum(p.size_usd for p in self.positions if p.category == category)

    def has_position(self, condition_id: str) -> bool:
        return any(p.condition_id == condition_id for p in self.positions)

    # ── Signal generation ─────────────────────────────────────────────

    def generate_signal(self, market: MarketInfo, estimate: Estimate) -> Optional[Signal]:
        """Compare estimate to market price, return Signal if edge exceeds threshold."""
        fair = estimate.fair_probability

        yes_edge = fair - market.outcome_yes_price
        no_edge = (1.0 - fair) - market.outcome_no_price

        if yes_edge > no_edge and yes_edge > self.config.min_edge:
            side = Side.YES
            edge = yes_edge
            market_price = market.outcome_yes_price
        elif no_edge > self.config.min_edge:
            side = Side.NO
            edge = no_edge
            market_price = market.outcome_no_price
        else:
            return None

        if market_price <= 0 or market_price >= 1:
            return None

        # Kelly criterion: f* = (b*p - q) / b
        b = (1.0 / market_price) - 1.0  # decimal odds
        p = fair if side == Side.YES else (1.0 - fair)
        q = 1.0 - p
        kelly_raw = (b * p - q) / b if b > 0 else 0.0
        kelly_raw = max(0.0, kelly_raw)

        # Fractional Kelly + position cap (use portfolio value, not just cash)
        kelly = kelly_raw * self.config.kelly_fraction
        portfolio_val = self.bankroll + self.total_exposure()
        size_usd = kelly * portfolio_val
        size_usd = min(size_usd, portfolio_val * self.config.max_position_pct)
        size_usd = min(size_usd, self.bankroll)  # never exceed available cash

        if size_usd < self.config.min_trade_usd:
            return None

        # CLOB minimum order size is 5 tokens → minimum USD = 5 * price
        min_clob_usd = 5.0 * market_price
        if size_usd < min_clob_usd:
            log.debug(f"Position ${size_usd:.2f} below CLOB minimum ${min_clob_usd:.2f} (5 tokens @ {market_price})")
            return None

        return Signal(
            market=market,
            estimate=estimate,
            side=side,
            edge=edge,
            market_price=market_price,
            kelly_fraction=kelly,
            position_size_usd=round(size_usd, 2),
            expected_value=round(size_usd * edge, 4),
        )

    # ── Risk checks ───────────────────────────────────────────────────

    def check_risk(self, signal: Signal) -> bool:
        """Return True if the trade passes all risk limits."""
        if self.has_position(signal.market.condition_id):
            log.info(f"Risk BLOCK: already positioned in {signal.market.question[:40]}")
            return False

        if len(self.positions) >= self.config.max_concurrent_positions:
            log.info(f"Risk BLOCK: max positions ({self.config.max_concurrent_positions}) reached")
            return False

        pv = self.bankroll + self.total_exposure()
        new_exposure = self.total_exposure() + signal.position_size_usd
        max_allowed = pv * self.config.max_total_exposure_pct
        if new_exposure > max_allowed:
            log.info(f"Risk BLOCK: total exposure ${new_exposure:.2f} > limit ${max_allowed:.2f}")
            return False

        cat_exp = self.category_exposure(signal.market.category) + signal.position_size_usd
        cat_limit = pv * self.config.max_category_exposure_pct
        if cat_exp > cat_limit:
            log.info(f"Risk BLOCK: '{signal.market.category}' exposure ${cat_exp:.2f} > limit ${cat_limit:.2f}")
            return False

        # Daily stop loss (include open position value — deployed capital isn't lost)
        portfolio_value = self.bankroll + self.total_exposure()
        daily_pnl = portfolio_value - self.daily_start_value
        if daily_pnl < 0 and abs(daily_pnl) > self.daily_start_value * self.config.daily_stop_loss_pct:
            log.warning(f"HALT: Daily stop loss triggered (PnL=${daily_pnl:+.2f}, limit={self.config.daily_stop_loss_pct:.0%})")
            self.is_halted = True
            return False

        # Max drawdown from high water mark
        if self.high_water_mark > 0:
            drawdown = (self.high_water_mark - portfolio_value) / self.high_water_mark
            if drawdown > self.config.max_drawdown_pct:
                log.warning(f"HALT: Max drawdown {drawdown:.1%} exceeded (limit={self.config.max_drawdown_pct:.0%})")
                self.is_halted = True
                return False

        # Agent death — only when total portfolio value (free cash + open positions)
        # drops below $1. Negative bankroll from API costs while holding positions
        # is normal: positions will eventually resolve and return USDC.
        if self.bankroll + self.total_exposure() < 1.0:
            log.warning("HALT: Portfolio value < $1 — agent is dead")
            self.is_halted = True
            return False

        return True

    # ── Position management ───────────────────────────────────────────

    def open_position(self, position: Position) -> None:
        """Deduct cost from bankroll and track position."""
        self.bankroll -= position.size_usd
        self.total_trades += 1
        self.positions.append(position)
        log.info(
            f"Opened {position.side.value} on {position.question[:40]}... "
            f"${position.size_usd:.2f} @ {position.entry_price:.3f}"
        )

    def close_position(self, condition_id: str, exit_price: float) -> float:
        """Close position, return cost basis + PnL to bankroll. Returns realized PnL."""
        pos = next((p for p in self.positions if p.condition_id == condition_id), None)
        if pos is None:
            return 0.0

        pnl = pos.shares * (exit_price - pos.entry_price)
        self.bankroll += pos.size_usd + pnl
        self.total_realized_pnl += pnl
        self.positions = [p for p in self.positions if p.condition_id != condition_id]
        self.high_water_mark = max(self.high_water_mark, self.bankroll)

        log.info(f"Closed {pos.question[:40]}... PnL: ${pnl:+.2f}")
        return pnl

    def resolve_position(self, condition_id: str, won: bool) -> float:
        """Close a resolved market position. Won: payout = shares * $1.00. Lost: payout = $0.
        Returns realized PnL."""
        pos = next((p for p in self.positions if p.condition_id == condition_id), None)
        if pos is None:
            return 0.0

        payout = pos.shares if won else 0.0
        pnl = payout - pos.size_usd
        self.bankroll += payout
        self.total_realized_pnl += pnl
        self.total_trades += 1
        self.positions = [p for p in self.positions if p.condition_id != condition_id]
        self.high_water_mark = max(self.high_water_mark, self.bankroll)

        result = "WON" if won else "LOST"
        log.info(f"Resolved ({result}): {pos.question[:40]}... payout=${payout:.2f}, PnL=${pnl:+.2f}")
        return pnl

    # ── Position review ─────────────────────────────────────────────

    def update_position_prices(self, prices: dict[str, float]) -> None:
        """Update current_price and unrealized_pnl for all positions with fresh market data."""
        for pos in self.positions:
            if pos.token_id in prices:
                pos.current_price = prices[pos.token_id]
                pos.unrealized_pnl = pos.shares * (pos.current_price - pos.entry_price)

    def generate_exit_signals(self) -> list[ExitSignal]:
        """Tier 1: free rule-based exit checks on all positions."""
        signals = []
        for pos in self.positions:
            # Skip unsellable positions: penny prices or below CLOB minimum (5 tokens)
            if pos.current_price < 0.01:
                log.debug(f"Skip review: {pos.question[:40]}... (price {pos.current_price:.4f} < $0.01)")
                continue
            if pos.shares < 5.0:
                log.debug(f"Skip review: {pos.question[:40]}... ({pos.shares:.2f} tokens < 5 minimum)")
                continue

            pnl = pos.shares * (pos.current_price - pos.entry_price)
            pnl_pct = (pos.current_price - pos.entry_price) / pos.entry_price if pos.entry_price > 0 else 0.0

            # Stop-loss: price dropped too far from entry
            if pnl_pct < -self.config.position_stop_loss_pct:
                signals.append(ExitSignal(pos, "stop_loss", pos.current_price, pnl, pnl_pct))
                continue

            # Take-profit: price near certainty
            if pos.current_price >= self.config.take_profit_price:
                signals.append(ExitSignal(pos, "take_profit", pos.current_price, pnl, pnl_pct))
                continue

            # Edge-gone: market moved past our original fair estimate
            if pos.fair_estimate_at_entry > 0:
                # fair_for_side: what we estimated the correct price for our side was
                fair_for_side = pos.fair_estimate_at_entry if pos.side == Side.YES else (1.0 - pos.fair_estimate_at_entry)
                if pos.current_price > fair_for_side + self.config.exit_edge_buffer:
                    signals.append(ExitSignal(pos, "edge_gone", pos.current_price, pnl, pnl_pct))
                    continue

        return signals

    def generate_topup_candidates(self) -> list[TopupCandidate]:
        """Tiny positions (<5 tokens) that want to exit — need a top-up BUY to reach CLOB minimum."""
        candidates = []
        for pos in self.positions:
            if pos.current_price < 0.01:
                continue  # penny = unsellable even with top-up
            if pos.shares >= 5.0:
                continue  # can sell normally (handled by generate_exit_signals)

            if pos.entry_price <= 0:
                continue

            pnl_pct = (pos.current_price - pos.entry_price) / pos.entry_price

            # Check same exit conditions as generate_exit_signals
            exit_reason = None
            if pnl_pct < -self.config.position_stop_loss_pct:
                exit_reason = "stop_loss"
            elif pos.current_price >= self.config.take_profit_price:
                exit_reason = "take_profit"
            elif pos.fair_estimate_at_entry > 0:
                fair_for_side = pos.fair_estimate_at_entry if pos.side == Side.YES else (1.0 - pos.fair_estimate_at_entry)
                if pos.current_price > fair_for_side + self.config.exit_edge_buffer:
                    exit_reason = "edge_gone"

            if exit_reason is None:
                continue

            topup_cost = 5.0 * pos.current_price
            recovery_value = pos.shares * pos.current_price

            candidates.append(TopupCandidate(
                position=pos,
                exit_reason=exit_reason,
                tokens_to_buy=5.0,
                topup_cost=topup_cost,
                recovery_value=recovery_value,
            ))

        return candidates

    def add_to_position(self, condition_id: str, additional_shares: float, additional_cost: float) -> None:
        """Add tokens to an existing position (used for top-up before sell)."""
        pos = next((p for p in self.positions if p.condition_id == condition_id), None)
        if pos is None:
            return
        pos.shares += additional_shares
        pos.size_usd += additional_cost
        self.bankroll -= additional_cost
        log.info(
            f"Top-up: +{additional_shares:.2f} tokens (${additional_cost:.2f}) -> "
            f"{pos.question[:40]}... now {pos.shares:.2f} tokens"
        )

    def get_review_candidates(self) -> list[Position]:
        """Tier 2: positions that moved significantly and should be re-estimated by Claude."""
        candidates = []
        for pos in self.positions:
            if pos.entry_price <= 0:
                continue
            price_move = abs(pos.current_price - pos.entry_price) / pos.entry_price
            if price_move >= self.config.review_reestimate_threshold_pct:
                candidates.append(pos)
        # Review biggest positions first
        candidates.sort(key=lambda p: p.size_usd, reverse=True)
        return candidates

    # ── Balance sync ─────────────────────────────────────────────────

    def sync_balance(self, actual_usdc_balance: float) -> None:
        """Sync bankroll from actual on-chain USDC balance.

        On-chain USDC is always the free cash (bankroll) — conditional tokens
        are held separately in the CLOB and not reflected in USDC balance.
        Always syncs both up and down so the bot has an accurate view of
        spendable funds (handles resolved-position payouts, deposits, fees, etc.)
        """
        prev = self.bankroll
        diff = actual_usdc_balance - prev
        if abs(diff) <= 0.001:
            return
        self.bankroll = actual_usdc_balance
        if diff > 0:
            log.info(
                f"Balance sync (upward): ${prev:.2f} -> ${self.bankroll:.2f} "
                f"(+${diff:.2f}, {len(self.positions)} positions open)"
            )
        else:
            log.warning(
                f"Balance sync (downward): ${prev:.2f} -> ${self.bankroll:.2f} "
                f"(${diff:.2f}, {len(self.positions)} positions open)"
            )
        self.high_water_mark = max(self.high_water_mark, self.bankroll + self.total_exposure())

    # ── Cost tracking ─────────────────────────────────────────────────

    def record_api_cost(self, input_tokens: int, output_tokens: int) -> None:
        """Deduct inference cost from bankroll. The agent pays for its own thinking."""
        cost = (input_tokens * INPUT_COST_PER_MTOK / 1_000_000) + \
               (output_tokens * OUTPUT_COST_PER_MTOK / 1_000_000)
        self.total_api_cost += cost
        self.bankroll -= cost

    def reset_daily(self) -> None:
        """Reset daily tracking. Call at the start of each new trading day."""
        self.daily_start_value = self.bankroll
